backend/server/src/http.rs

```diff
@@ -53,77 +63,117 @@ pub async fn serve(
-            //
-            // TODO: do we want to remove this route in production?
-            (&Method::GET, "/graphiql") => Box::pin(juniper_hyper::graphiql("/graphql", None)),
-
-            // The actual GraphQL API.
-            (&Method::GET, "/graphql") | (&Method::POST, "/graphql") => {
-                let result = juniper_hyper::graphql(
-                    self.root_node.clone(),
-                    self.context.clone(),
-                    req,
-                );
-
-                Box::pin(result)
-            }
+/// This is the main entry point, called for each incoming request.
+async fn handle(req: Request<Body>, ctx: Arc<Context>) -> Result<Response, hyper::Error> {
+    trace!(
+        "Incoming HTTP {:?} request to '{}{}'",
+        req.method(),
+        req.uri().path(),
+        req.uri().query().map(|q| format!("?{}", q)).unwrap_or_default(),
+    );
 
-            // Realm pages
-            (&Method::GET, path) if path.starts_with(REALM_PREFIX) => {
-                let _realm_path = &path[REALM_PREFIX.len()..];
-                // TODO: check if path is valid
+    let method = req.method();
+    let path = req.uri().path();
 
-                serve_index()
-            }
+    const ASSET_PREFIX: &str = "/assets/";
+    const REALM_PREFIX: &str = "/r/";
 
-            // Assets (JS files, fonts, ...)
-            (&Method::GET, path) if path.starts_with(ASSET_PREFIX) => {
-                let asset_path = &path[ASSET_PREFIX.len()..];
-                Assets::serve(asset_path).unwrap_or_else(|| reply_404(method, path))
-            }
+    let response = match (method, path) {
+        (&Method::GET, "/") => serve_index().await,
+
+        // The interactive GraphQL API explorer/IDE. We actually keep this in
+        // production as it does not hurt and in particular: does not expose any
+        // information that isn't already exposed by the API itself.
+        (&Method::GET, "/graphiql") => juniper_hyper::graphiql("/graphql", None).await?,
 
-            // 404 for everything else
-            (method, path) => reply_404(method, path),
+        // The actual GraphQL API.
+        (&Method::GET, "/graphql") | (&Method::POST, "/graphql") => {
+            juniper_hyper::graphql(ctx.api_root.clone(), ctx.api_context.clone(), req).await?
+        }
+
+        // Realm pages
+        (&Method::GET, path) if path.starts_with(REALM_PREFIX) => {
+            let _realm_path = &path[REALM_PREFIX.len()..];
+            // TODO: check if path is valid
+
+            serve_index().await
         }
-    }
-}
 
-type ResponseFuture = <RootService as Service<Request<Body>>>::Future;
+        // Assets (JS files, fonts, ...)
+        (&Method::GET, path) if path.starts_with(ASSET_PREFIX) => {
+            let asset_path = &path[ASSET_PREFIX.len()..];
+            match Assets::serve(asset_path).await {
+                Some(r) => r,
+                None => reply_404(method, path).await,
+            }
```

Comment:
It's a sad story :/ 

So `Assets::serve` is an async fn returning an `Option`. This means it returns `impl Future<Output = Option<T>>`. So we have to await it to get an `Option<T>`. You can't write `.unwrap_or(|| reply_404(...))` because `reply_404` is async as well. So you gotta `.await` it, but you can't await in the closure of `unwrap_or` :/ One could define `Assets::serve` to manually return `Option<impl Future<Output = T>>` (and remove the `async`), but that's also kinda meh and uhg.

Reply:
:cry: